### ISSUE TITLE
[Bug] Invalid staging when "ignore whitespaces" is on

### ISSUE THREAD

**Issue by sm-g** [ForkIssues/TrackerWin#202]
Repro is tricky, because I cannot copypaste source here. 

Crete changes with some lines only different in spacing.
Enable "ignore whitespaces".
Selection on hovering Stage button looks like we stage only visible changes.

But when you click Stage, you either actually stage changes with spaces also, or, if you are lucky, got 
```
error: patch failed: SourceCode/xxx.cs:288
error: SourceCode/xxx.cs: patch does not apply```

**Comment by DanPristupov** [comment]
There's nothing I can do. Sometimes the given context is not enough for git to recognize the chunk.

Generally speaking, 'ignore whitespaces' should be used for review rather than for staging.

**Comment by sm-g** [comment]
Better disable staging by chunks then

**Comment by DanPristupov** [comment]
Maybe, but in this way you would file another issue.

**Comment by chucker** [comment]
> There's nothing I can do. Sometimes the given context is not enough for git to recognize the chunk.

Both GitKraken and Sourcetree succeed with the same hunk. My guess is they apply some fuzzy logic.

**Comment by DanPristupov** [comment]
> Both GitKraken and Sourcetree succeed with the same hunk. 

Do you have 'ignore whitespaces' ignored in both of them?
Do you remember if the number of context lines were equal? Adding more lines usually helps.

**Comment by carlolars** [comment]
I get the same error from time to time. Increasing the number of context lines does not help but disabling *ignore whitespaces* works.

**Comment by carlolars** [comment]
Got the error again now.  
Applying the same patch from the command line cause the same error when the diff is created with `--ignore-space-change`. Applying the patch using a diff created without `--ignore-space-change` is working.  
But, if adding the option `--ignore-whitespace` to `git apply` then adding the diff with `--ignore-space-change` is working, so maybe add that option when staging chunks and _ignore-whitespace_ is enabled?

The error can be reproduced using the following simple example:
Original committed code:
```c
void func(int x)
{
    printf("hello");
}
```
Modification that fails when adding chunk with ignore-whitespace enabled:
```c
void func(int x)
{
    if(x != 0)
    {
        printf("hello");
    }
}
```

**Comment by DanPristupov** [comment]
@carlolars 

I was mistakingly sure I already do that. However I just checked the source code and it turned out that I do `git apply --cached --whitespace=nowarn` which is an absolute BS [(docs)](https://git-scm.com/docs/git-apply#Documentation/git-apply.txt---whitespaceltactiongt). I suppose this is a 4 years old mistake which has migrated from Fork for Mac and has been existing since Fork 1.0.2 or so. I reviewed this code a lot of times and was always sure that `--whitespace=nowarn` ignores the whitespace changes :).

Right now I'm in the middle of a large task, so I will try what you said a bit later and share the results.
